arangod/Cluster/AgencyCache.cpp

```diff
@@ -315,111 +299,118 @@ void AgencyCache::run() {
-                          _planChanges.emplace(_commitIndex, std::move(i));
-                        }
-                        for (auto const& i : cc) {
-                          _currentChanges.emplace(_commitIndex, std::move(i));
-                        }
-                      }
+                TRI_ASSERT(rs.hasKey("log"));
+                TRI_ASSERT(rs.get("log").isArray());
+                LOG_TOPIC("4579e", TRACE, Logger::CLUSTER) <<
+                  "Applying to cache " << rs.get("log").toJson();
+                std::unordered_set<std::string> pc;  // Plan changes
+                std::unordered_set<std::string> cc;  // Current changes
+                for (auto const& i : VPackArrayIterator(rs.get("log"))) {
+                  pc.clear();
+                  cc.clear();
+                  {
+                    std::lock_guard g(_storeLock);
+                    _readDB.applyTransaction(i); // apply logs
+                    _commitIndex = i.get("index").getNumber<uint64_t>();
+
+                    {
+                      std::lock_guard g(_callbacksLock);
+                      handleCallbacksNoLock(i.get("query"), uniq, toCall, pc, cc);
+                    }
+
+                    for (auto const& i : pc) {
+                      _planChanges.emplace(_commitIndex, i);
+                    }
+                    for (auto const& i : cc) {
+                      _currentChanges.emplace(_commitIndex, i);
                     }
                   }
-                } else {
-                  TRI_ASSERT(rs.hasKey("readDB"));
-                  std::lock_guard g(_storeLock);
-                  LOG_TOPIC("4579f", TRACE, Logger::CLUSTER) <<
-                    "Fresh start: overwriting agency cache with " << rs.toJson();
-                  _readDB = rs;                  // overwrite
-                  pc = reInitPlan();
-                  _initialized.store(true, std::memory_order_relaxed);
-                  for (auto const& i : pc) {
-                    _planChanges.emplace(_commitIndex, std::move(i));
-                  }
-                  // !! Check documentation of the function before making changes here !!
-                  _commitIndex = commitIndex;
-                  _lastSnapshot = commitIndex;
                 }
-                triggerWaiting(commitIndex);
-                if (firstIndex > 0) {
-                  if (!toCall.empty()) {
-                    invokeCallbacks(toCall);
-                  }
-                } else {
-                  invokeAllCallbacks();
+              } else {
+                // firstIndex == 0, we got a snapshot:
+                TRI_ASSERT(rs.hasKey("readDB"));
+                std::lock_guard g(_storeLock);
+                LOG_TOPIC("4579f", TRACE, Logger::CLUSTER) <<
+                  "Fresh start: overwriting agency cache with " << rs.toJson();
+                _readDB = rs;                  // overwrite
+                _initialized.store(true, std::memory_order_relaxed);
```

Comment:
I think we should move the setting of `_initialized` to `true` a bit further down, when it is known that the initialization was actually fully completed.
Otherwise it may not be safe if one of the following `emplace(...)` calls fails. In this case `_initialized` will be true but the `_planChanges` map is not complete.